ow_plexil/src/plexil-adapter/FaultDependencies.cpp

```diff
@@ -0,0 +1,355 @@
+    m_subsystems[parent_subsystem].faulty = 1; 
+  }
+  else{
+    m_subsystems[parent_subsystem].faulty = 0; 
+  }
+
+  for (auto dep: m_faults[name].impacts){
+    // If an impact to update is a subsystem
+    if (dep.second == "Subsystem"){
+      if (m_subsystems.find(dep.first) != m_subsystems.end()){
+        updateSubsystem(dep.first, status, parent_subsystem);
+      }
+    }
+    // Otherwise we treat it as a procedure to update
+    else{
+      if (m_procedures.find(dep.first) != m_procedures.end()){
+        if (status == 1){
+            m_procedures[dep.first].inoperable += 1;
+        }
+        else{
+            m_procedures[dep.first].inoperable -= 1;
+        }
+      }
+    }
+ 
+  }
+  // If verbose output flag is true we print the current fault dependency info
+  if (m_verbose_flag){
+    DebugPrint();
+  }
+}
+
+void FaultDependencies::updateSubsystem(const std::string &name, int status, const std::string &parent){
+  int prev_inoperable_state = m_subsystems[name].inoperable;
+  // increment non_local fault and inoperable flags if status is 1
+  if (status == 1){
+      if (parent != name){
+        m_subsystems[name].non_local_fault += 1;
+      }
+      m_subsystems[name].inoperable += 1;
+      ROS_INFO("SUBSYSTEM: %s INOPERABLE.", name.c_str());
+  }
+  // otherwise decrement
+  else{
+      if (parent != name){
+        m_subsystems[name].non_local_fault -= 1;
+      }
+      m_subsystems[name].inoperable -= 1;
+      if (m_subsystems[name].inoperable == 0){
+        ROS_INFO("SUBSYSTEM: %s NOW OPERABLE.", name.c_str());
+      }
+  }
+  // If subsystem inoperable flag changes, we need to cascade inoperable fault to impacted subsystems
+  if (prev_inoperable_state != m_subsystems[name].inoperable){
+    cascadeFault(name, status);
+  }
+}
+
+void FaultDependencies::cascadeFault(const std::string &subsystem_name, int status){
+  std::unordered_set<std::string> visited;
+  std::stack<std::string> nodes;
+
+  visited.insert(subsystem_name);
+  nodes.push(subsystem_name);
+  
+    // BFS through all subsystem impacts
+    while(!nodes.empty()){
+      std::string current_node = nodes.top();
+      nodes.pop();
+
+      for (auto i: m_subsystems[current_node].impacts){
+        // If it is a procedure we do not need to add it to BFS stack, and can just update the inoperable flag
+        if (i.second == "Procedure"){
+          m_procedures[i.first].inoperable = status;
+        }
+        // Otherwise we update subsystem and add it to stack
+        else{
+          if (visited.insert(i.first).second && m_subsystems.find(i.first) != m_subsystems.end()){
+            // increment or decrement non_local_fault flag
+            if (status){
+              m_subsystems[i.first].non_local_fault += 1;
+              m_subsystems[i.first].inoperable += 1;
+            }
+            else{
+              m_subsystems[i.first].non_local_fault -= 1;
+              m_subsystems[i.first].inoperable -= 1;
+            }
```

Comment:
This looks similar to enough to what you're doing at line 141. I say make it a function so that updating itself is the business of the subsystem.
```suggestion
            m_subsystems[i.first].update(status);
```

Reply:
They are very similar, but the one present in updateSubsystems has an extra if statement to see if the fault is local (which cascadeFault does not need).

I think breaking this into a function would make it harder to read, and require an additional argument for local fault checking.

Reply:
This is pretty minor, so I'm fine with it staying how it is.